### PR TITLE
Deal with memory leak caused by string and token

### PR DESCRIPTION
**Pull request by yetingk**
I let token lie on stack instead of heap, since I think we don't need token live too long.

### CODE REVIEW

**Comment by jserv** [comment]
The commit messages should append the memory leak comparisons between w/ and w/o this patch via valgrind.

**Comment by yetingk** [comment]
The difference between this modification when Inputing this command, 

valgrind --leak-check=full ./mzcc < sample/nqueen.c > sample/nqueen.s

Before modification:                                                                                                                                                                                            

==15301== 10,032 (24 direct, 10,008 indirect) bytes in 1 blocks are definitely lost in loss record 2,017 of 2,017
==15301==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15301==    by 0x404ED5: make_list (list.h:25)
==15301==    by 0x408626: read_toplevels (parser.c:1076)
==15301==    by 0x409DE6: main (main.c:9)
==15301== 
==15301== LEAK SUMMARY:
==15301==    definitely lost: 81,536 bytes in 3,557 blocks
==15301==    indirectly lost: 75,680 bytes in 1,147 blocks
==15301==      possibly lost: 0 bytes in 0 blocks
==15301==    still reachable: 544 bytes in 20 blocks
==15301==         suppressed: 0 bytes in 0 blocks

==15301== ERROR SUMMARY: 1447 errors from 1447 contexts (suppressed: 0 from 0)

After modification:

==14594== 10,032 (24 direct, 10,008 indirect) bytes in 1 blocks are definitely lost in loss record 1,050 of 1,050
==14594==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==14594==    by 0x4050AE: make_list (list.h:25)
==14594==    by 0x408A52: read_toplevels (parser.c:1072)
==14594==    by 0x40A209: main (main.c:9)
==14594== 
==14594== LEAK SUMMARY:
==14594==    definitely lost: 71,344 bytes in 1,568 blocks
==14594==    indirectly lost: 17,064 bytes in 549 blocks
==14594==      possibly lost: 0 bytes in 0 blocks
==14594==    still reachable: 544 bytes in 20 blocks
==14594==         suppressed: 0 bytes in 0 blocks

==15445== ERROR SUMMARY: 778 errors from 778 contexts (suppressed: 0 from 0)

**Comment by jserv** [comment]
> valgrind --leak-check=full ./mzcc < sample/nqueen.c > sample/nqueen.s
> Before modification:
> ==15301== definitely lost: 81,536 bytes in 3,557 blocks
> ==15301== indirectly lost: 75,680 bytes in 1,147 blocks
> After modification:
> ==14594== definitely lost: 71,344 bytes in 1,568 blocks
> ==14594== indirectly lost: 17,064 bytes in 549 blocks

The commit message was not accurate. This patch reduced memory leaks, but it still exists somewhere. We have to figure out the exact source.

File: verbose.c [hunk 1]
```diff
@@ -232,7 +233,8 @@ char *token_to_string(Token *tok)
     case TTYPE_STRING:
         string_appendf(s, "\"%s\"", get_strtok(tok));
         return get_cstring(s);
+    default:
+        error("internal error: unknown token type: %d", get_ttype(tok));
```
Comment (jserv) [inline verbose.c]: Move this statement out of `switch` scope.

File: util.h [hunk 2]
```diff
@@ -26,15 +26,16 @@ static inline String *make_string(void)
...
+    char *body = realloc(s->body, newsize);
     s->body = body;
     s->nalloc = newsize;
 }
 
 static inline char *get_cstring(String *s)
 {
-    return s->body;
+    char *ret = s->body;
+    free(s);
```
Comment (jserv) [inline util.h]: I am not sure if it necessary to call `free` here.

File: verbose.c (hunk 1 above)
Comment (fakepaper56) [reply verbose.c]: But if I mover this statement out of switch, the output by "make check": 
verbose.c:219:5: error: enumeration value ‘TTYPE_NULL’ not handled in switch [-Werror=switch]
     switch (ttype) {
     ^~~~~~

Comment (jserv) [reply verbose.c]: I misinterpreted the context.

File: util.h (hunk 2 above)
Comment (jserv) [reply util.h]: All getter functions should touch data fields or members without releasing resources accordingly.

**Comment by yetingk** [comment]
I use the same trick, struct initialization, on list iterator and strings

**Comment by jserv** [comment]
Use `git rebase -i` to rework each git commit message to contain the summary of Valgrind report.

File: codegen_x64.c [hunk 3]
```diff
@@ -765,8 +765,8 @@ static void emit_func_prologue(Ast *func)
         off -= align(v->ctype->size, 8);
         v->loff = off;
     }
-    for (Iter *i = list_iter(func->localvars); !iter_end(i);) {
-        Ast *v = iter_next(i);
+    for (Iter i = list_iter(func->localvars); !iter_end(i);) {
```
Comment (jserv) [inline codegen_x64.c:768]: By the way, I am thinking of the introduce for [for_each macro](https://lwn.net/Articles/336255/) used in Linux kernel to simplify the similar loops here. Let's do that in another pull request.

File: util.h [hunk 4]
```diff
@@ -13,28 +13,28 @@ typedef struct {
...
+static inline String make_string(void)
 {
-    String *r = malloc(sizeof(String));
-    r->body = malloc(INIT_SIZE);
-    r->nalloc = INIT_SIZE;
-    r->len = 0;
-    r->body[0] = '\0';
-    return r;
+    char *body = malloc(INIT_SIZE);
+    body[0] = '\0';
```
Comment (jserv) [inline util.h]: Replace `malloc` with `calloc` to ensure allocated memory filled with bytes of value zero.

File: list.h [hunk 5]
```diff
@@ -109,4 +109,17 @@ static inline int list_len(List *list)
...
+    ListNode *now = iter->ptr;
+    iter->ptr = now->next;
+    free(now->elem);
+    free(now);
+}
+
+static inline void list_free(List *list)
+{
+    for (Iter i = list_iter(list); !iter_end(i);)
+        iter_next_and_free(&i);
```
Comment (jserv) [inline list.h]: Safely move function body of `iter_next_and_free` here.
